### PR TITLE
init skeleton for scz bp cohort definition

### PR DESCRIPTION
**Pull request by HLasse**
# Notes for reviewers

### CODE REVIEW

File: psycop/common/cohort_definition.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class Cohort(ABC):
+    """Interface for a cohort definition."""
+
+    @abstractmethod
+    def get_eligible_prediction_times(self) -> FilteredCohort:
+        ...
+
+    @abstractmethod
+    def get_outcome_timestamps(self) -> pl.DataFrame:
+        ...
```
Comment (MartinBernstorff) [inline psycop/common/cohort_definition.py:38]: Turn these into static methods? :-)

File: psycop/common/cohort_definition.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import polars as pl
+from pydantic import BaseModel
+
+
+@dataclass
+class StepDelta:
+    step_name: str
+    n_before: int
+    n_after: int
```
Comment (MartinBernstorff) [inline psycop/common/cohort_definition.py:12]:
```suggestion
    n_after: int
    step_index: int
```

Currently, FilteredCohort has a list of StepDeltas. Whether this is a queue or a stack will determine the ordering. I'd love to have this in each StepDelta as well, to ensure that reordering the list doesn't screw up the ordering :-)

File: psycop/common/cohort_definition.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import polars as pl
+from pydantic import BaseModel
+
+
+@dataclass
```
Comment (MartinBernstorff) [inline psycop/common/cohort_definition.py:8]:
```suggestion
@dataclass(frozen=True)
```

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_eligible_config.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from datetime import datetime
+
+MIN_AGE = 18
+MAX_AGE = 40
+MIN_DATE = datetime(year=2013, month=1, day=1)
+N_DAYS_WASHIN = 182
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_eligible_config.py:6]: Is this tested? I.e. is there a flattening in the incidence after this point?

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    get_first_scz_bp_diagnosis_after_washin,
+)
+
+
+class SczBpCohort(Cohort):
+    def __init__(self):
+        self.stepdeltas: list[StepDelta] = []
+
+    def get_eligible_prediction_times(self) -> FilteredCohort:
+        prediction_times = self._get_eligible_prediction_times()
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py:26]: What is the advantage of this nesting? :-)

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            self.stepdeltas.append(
+                StepDelta(
+                    step_name=filtering_step_name,
+                    n_before=n_before,
+                    n_after=df.shape[0],
+                ),
+            )
+        return df
+
+    def _get_filtering_steps(self) -> dict[str, Callable]:
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py:58]: We might want to type hint Callables, sets etc. E.g. if it both takes and returns a pl.DataFrame, `Callable[pl.Dataframe, pl.Dataframe]`. 

Alternatively, perhaps an ABC for filtering steps?

File: psycop/common/cohort_definition.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    @property
+    def n_dropped(self) -> int:
+        return self.n_before - self.n_after
+
+
+class FilteredCohort(BaseModel):
+    """A filtered cohort is a cohort that has been filtered by a set of criteria."""
+
+    cohort: pl.DataFrame
```
Comment (MartinBernstorff) [inline psycop/common/cohort_definition.py:22]: Might want to represent the granularity here (e.g. this is not at the level of patients). For example:
```suggestion
    prediction_times: pl.DataFrame
```

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    without_prevalent_scz_or_bp,
+)
+from psycop.projects.scz_bp.feature_generation.outcome_specification.first_scz_or_bp_diagnosis import (
+    get_first_scz_bp_diagnosis_after_washin,
+)
+
+
+class SczBpCohort(Cohort):
+    def __init__(self):
+        self.stepdeltas: list[StepDelta] = []
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py:23]: We collect the stepdeltas in two places now - in FilteredCohort and in self.stepdeltas. I would make this private, and encourage using from FilteredCohort.


```suggestion
        self.stepdeltas: list[StepDelta] = []
```

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            "min_date": min_date,
+            "washin": excluded_by_washin,
+            "washout_move": washout_move,
+            "prevalent_scz_or_bp": without_prevalent_scz_or_bp,
+        }
+
+
+if __name__ == "__main__":
+    cohort = SczBpCohort()
+    df = cohort.get_eligible_prediction_times().cohort
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py:72]: If we use the FilteredCohort object, this should be:

cohort = cohort_generator.get_eligible_prediction_times()

File: psycop/common/cohort_definition.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    filter_steps: list[StepDelta]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class Cohort(ABC):
```
Comment (MartinBernstorff) [inline psycop/common/cohort_definition.py:29]: Cohort definer

File: psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    for df_type in dfs:
+        dfs[df_type]["source"] = df_type
+
+    combined = pd.concat(
+        [dfs[k] for k in dfs],
+        axis=0,
+    )
+
+    first_scz_or_bp_indicator = (
+        combined.sort_values("timestamp")
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py:29]: Have you tested whether this is ascending or descending? :-)

File: psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        .groupby("dw_ek_borger")
+        .first()
+        .reset_index(drop=False)
+    )
+    # subtract 1 day from first diagnosis to avoid any leakage and add indicator column
+    first_scz_or_bp_indicator = pl.from_pandas(first_scz_or_bp_indicator).with_columns(
+        pl.col("timestamp") - pl.duration(days=1),
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py:36]: I assume this overrides the "timestamp" column, tested?

File: psycop/projects/scz_bp/feature_generation/outcome_specification/add_time_from_first_visit.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+                return_value_as_visit_length_days=False,
+                timestamp_for_output="start",
+            ),
+        )
+        .groupby("dw_ek_borger")
+        .agg(pl.col("timestamp").min().alias("first_visit"))
+    )
+    df = df.join(first_visit, on="dw_ek_borger", how="left")
+
+    df = df.with_columns(
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/outcome_specification/add_time_from_first_visit.py:23]: Bare som en neat ting, så kan du bruge pl.col("timestamp").min().over("dw_ek_borger).alias("first_visit") til at erstatte hele blokken her :-)

https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.over.html

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
+from typing import Callable
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py:1]: Rename file to prediction_time_loader to disambiguate?

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py (hunk 6 above)
Comment (HLasse) [reply psycop/projects/scz_bp/feature_generation/eligible_prediction_times/scz_bp_cohort_loader.py:58]: Yeah, considered making a common interface for filtering steps too. I think it's a bit too much boilerplate and that typehinting the Callable would be a good alternative

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/single_filters.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    time_of_first_scz_bp_diagnosis = get_first_scz_or_bp_diagnosis().select(
+        pl.col("timestamp").alias("timestamp_outcome"),
+        pl.col("dw_ek_borger"),
+    )
+
+    prediction_times_with_outcome = df.filter(
+        pl.col("dw_ek_borger").is_in(
+            time_of_first_scz_bp_diagnosis.get_column("dw_ek_borger"),
+        ),
+    ).join(time_of_first_scz_bp_diagnosis, on="dw_ek_borger", how="inner")
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/single_filters.py:69]: Får vi ikke den filtrering automatisk når der laves et inner_join nedenfor? :-)

File: psycop/projects/scz_bp/feature_generation/eligible_prediction_times/single_filters.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+    prevalent_prediction_times = prediction_times_with_outcome.filter(
+        pl.col("timestamp") > pl.col("timestamp_outcome"),
+    )
+
+    return df.join(prevalent_prediction_times, on="dw_ek_borger", how="anti")
+
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/eligible_prediction_times/single_filters.py:76]: Her tror jeg du fjerner alle patienter, der har mindst en prævalent prediction time. Men du vil nok kun fjerne de prævalente prediction times, ikke alle prævalente patienter, right?

Måske en fordel hvis det fremgår i funktionsnavnet, om filtreringen skal ske på patient- eller prediction time niveau :-)

File: psycop/common/feature_generation/loaders/raw/load_scz_bp_outcomes.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+from typing import Literal
+
+import pandas as pd
+
+from psycop.common.feature_generation.loaders.raw.load_diagnoses import from_contacts
+from psycop.common.feature_generation.utils import data_loaders
+
+
+@data_loaders.register("manic_bipolar_outcome")
+def manic_and_bipolar(
```
Comment (MartinBernstorff) [inline psycop/common/feature_generation/loaders/raw/load_scz_bp_outcomes.py:11]:
```suggestion
def manic_or_bipolar(
```

File: psycop/common/feature_generation/loaders/raw/load_scz_bp_outcomes.py [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+        n_rows=n_rows,
+        shak_location_col=shak_location_col,
+        shak_code=shak_code,
+        shak_sql_operator=shak_sql_operator,
+        timestamp_purpose=timestamp_purpose,
+    )
+
+
+@data_loaders.register("scz_outcome")
+def scz_and_sczaffective(
```
Comment (MartinBernstorff) [inline psycop/common/feature_generation/loaders/raw/load_scz_bp_outcomes.py:30]:
```suggestion
def scz_or_sczaffective(
```

File: psycop/projects/scz_bp/feature_generation/outcome_specification/outcome_eda.Rmd [hunk 19]
```diff
@@ -0,0 +1,288 @@
+---
+title: "kohorte_udtræk"
+output: html_document
+---
+
+Dette script udtrækker kohorten til prædiktion af bipolar lidelse og skizofrenis. Falder i to dele
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/outcome_specification/outcome_eda.Rmd:6]: Har ikke reviewet her. Hvis du er afhængig af det i senere analyser, så giv lyd :-)

File: psycop/projects/scz_bp/feature_generation/outcome_specification/scz_diagnoses.py [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+from psycop.common.feature_generation.loaders.raw.load_scz_bp_outcomes import (
+    scz_and_sczaffective,
+)
+
+
+def get_first_scz_diagnosis() -> pd.DataFrame:
+    df = scz_and_sczaffective()
+
+    df_first_scz_diagnosis = (
+        df.sort_values("timestamp")
```
Comment (MartinBernstorff) [inline psycop/projects/scz_bp/feature_generation/outcome_specification/scz_diagnoses.py:12]: Marker om ascending eller descending.

**Review by MartinBernstorff — COMMENTED** [review]
Great first stab! A handful of suggestions below, lemme know what you think 👍

File: psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py (hunk 11 above)
Comment (HLasse) [reply psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py:29]: Defaults to ascending - added to be explicit :+1:

File: psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py (hunk 12 above)
Comment (HLasse) [reply psycop/projects/scz_bp/feature_generation/outcome_specification/first_scz_or_bp_diagnosis.py:36]: Yep!

File: psycop/projects/scz_bp/feature_generation/outcome_specification/add_time_from_first_visit.py (hunk 13 above)
Comment (HLasse) [reply psycop/projects/scz_bp/feature_generation/outcome_specification/add_time_from_first_visit.py:23]: Neat pattern! Her har jeg dog brug for aggregeringen (fra mange besøg til kun det første) - men tak for obs!

**Review by MartinBernstorff — APPROVED** [review]

**Pull request merged**